source/solvers/navier_stokes_base.cc

```diff
@@ -1188,6 +1188,11 @@ template <int dim, typename VectorType, typename DofsType>
 void
 NavierStokesBase<dim, VectorType, DofsType>::refine_mesh_uniform()
 {
+  auto &tria = *dynamic_cast<parallel::distributed::Triangulation<dim> *>(
+    this->triangulation.get());
+  if (tria.n_levels() >
+      this->simulation_parameters.mesh_adaptation.maximum_refinement_level)
+    return;
```

Comment:
I don't think you need this dynamic cast here.
The very base triangulation class: https://dealii.org/current/doxygen/deal.II/classTriangulation.html#a777f035a17e91a4d822971516ca11db5

Has this function n_levels.  ( see [n_levels](https://dealii.org/current/doxygen/deal.II/classTriangulation.html#a777f035a17e91a4d822971516ca11db5))

So I don't think you need to case before this refinement.
For sure we don't support refining simplex meshes, so that could be asserted to prevent a bug ( I don't know what happens if you try to refine a simplex mesh now)